chain/consensus/mir/consensus.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/filecoin-project/lotus/chain/actors/builtin/reward"
+	"github.com/filecoin-project/lotus/chain/beacon"
+	"github.com/filecoin-project/lotus/chain/consensus"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+)
+
+var _ consensus.Consensus = &Mir{}
+
+var RewardFunc = func(ctx context.Context, vmi vm.Interface, em stmgr.ExecMonitor,
+	epoch abi.ChainEpoch, ts *types.TipSet, params *reward.AwardBlockRewardParams) error {
+	// TODO: No RewardFunc implemented for mir yet
+	return nil
+}
+
+type Mir struct {
+	beacon  beacon.Schedule
+	sm      *stmgr.StateManager
+	genesis *types.TipSet
+}
+
+func NewConsensus(
+	ctx context.Context,
+	sm *stmgr.StateManager,
+	b beacon.Schedule,
+	g chain.Genesis,
+	netName dtypes.NetworkName,
+) (consensus.Consensus, error) {
+	return &Mir{
+		beacon:  b,
+		sm:      sm,
+		genesis: g,
+	}, nil
+}
+
+// CreateBlock creates a Filecoin block from the block template provided by Mir.
+func (bft *Mir) CreateBlock(ctx context.Context, w lapi.Wallet, bt *lapi.BlockTemplate) (*types.FullBlock, error) {
+	pts, err := bft.sm.ChainStore().LoadTipSet(ctx, bt.Parents)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load parent tipset: %w", err)
+	}
+
+	st, recpts, err := bft.sm.TipSetState(ctx, pts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load tipset state: %w", err)
+	}
+
+	next := &types.BlockHeader{
+		Miner:         builtin.SystemActorAddr, // Mir blocks are not signed, we use system addr as miner.
+		Parents:       bt.Parents.Cids(),
+		Ticket:        bt.Ticket,
+		ElectionProof: bt.Eproof,
+
+		BeaconEntries: bt.BeaconValues,
+		Height:        bt.Epoch,
+		// Each validator in Mir be assembling the block with a different
+		// timestamp. To avoid validators from pushing blocks with different
+		// timestamps that lead to different CIDs, we use the epoch as
+		// a timestamp for now.
+		// TODO: Consider exporting a batch timestamp from Mir and use it
+		// for the block timestamp.
+		Timestamp:             uint64(bt.Epoch),
+		WinPoStProof:          bt.WinningPoStProof,
+		ParentStateRoot:       st,
+		ParentMessageReceipts: recpts,
+	}
+	blsMessages, secpkMessages, err := consensus.MsgsFromBlockTemplate(ctx, bft.sm, next, pts, bt)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to process messages from block template: %w", err)
+	}
+
+	return &types.FullBlock{
+		Header:        next,
+		BlsMessages:   blsMessages,
+		SecpkMessages: secpkMessages,
+	}, nil
+}
+
+func (bft *Mir) ValidateBlockHeader(ctx context.Context, b *types.BlockHeader) (rejectReason string, err error) {
+	// TODO: This is missing.
+	log.Warn("oh oh! No specific block header validation implemented for Mir yet")
+	return "", nil
+}
+
+func (bft *Mir) ValidateBlock(ctx context.Context, b *types.FullBlock) (err error) {
```

Comment:
This PR adds only support for Mir validators. Validators can only receive blocks through the orderer layer, and they don't broadcast blocks to the network. Support for full-nodes is introduced in the next [PR](https://github.com/consensus-shipyard/lotus/pull/19) (maybe it would have made sense to merge both, but it was getting huge).

The relevant code for block validation can be found here: https://github.com/consensus-shipyard/lotus/blob/510c6dfc77485aa463cde07a9f9bccee86addc87/chain/consensus/mir/consensus.go#L106

Sorry if we made this a mess, happy to merge PRs and make it a single on, and/or walk you through the code if it helps :pray: